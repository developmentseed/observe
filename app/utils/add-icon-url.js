import * as taginfo from '../presets/taginfo.json'
import { uninterestingKeys } from './uninterestingKeys'
import _omit from 'lodash.omit'
import _find from 'lodash.find'

/**
 * Parses through each feature, and adds icon url to properties
 * The icon property is used for styling later
 * @param {Object} geojson
 * @return {Object} geojson
 */
export function addIconUrl (geojson) {
  if (geojson.type === 'FeatureCollection') {
    const points = geojson.features.filter(feature => {
      return feature.geometry.type === 'Point'
    })
    points.forEach(feature => {
      setIcon(feature)
    })
  } else {
    setIcon(geojson)
  }
  return geojson
}

function setIcon (feature) {
  feature.properties.icon = undefined
  const featureTags = _omit(feature.properties, uninterestingKeys)
  const keys = Object.keys(featureTags)
  for (const index in keys) {
    const key = keys[index]
    const match = _find(taginfo.tags, { 'key': key, 'value': featureTags[key] })
    if (match) {
      if (match.icon_url != null) {
        feature.properties.icon = match.icon_url.replace(/-/g, '_')
      }
      break
    }
  }
  if (!feature.properties.icon || feature.properties.icon.startsWith('iD-')) {
    feature.properties.icon = 'maki_marker'
  }
  return feature
}
