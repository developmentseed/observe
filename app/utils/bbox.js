import tileCover from '@mapbox/tile-cover'
import bboxPolygon from '@turf/bbox-polygon'
import turfBbox from '@turf/bbox'
import turfBuffer from '@turf/buffer'
import _flatten from 'lodash.flatten'

// TODO - move this to an appropriate config file
const tileLimits = {
  min_zoom: 16,
  max_zoom: 16
}

/**
 * Convert map bbox object to a string
 * @param {*} bbox
 * @return string
 */
export function bboxToString (bbox) {
  const ne = bbox[0]
  const sw = bbox[1]
  const left = sw[0]
  const bottom = sw[1]
  const right = ne[0]
  const top = ne[1]
  return [left, bottom, right, top].join(',')
}

export function bboxToTiles (bbox) {
  const polygon = bboxPolygon(_flatten(bbox))
  const tiles = tileCover.indexes(polygon.geometry, tileLimits)
  return tiles
}

export function featureToTiles (feature) {
  let bbox = []
  if (feature.geometry && feature.geometry.type === 'Point') {
    const buffer = turfBuffer(feature, '0.0005', { 'units': 'kilometers' })
    bbox = turfBbox(buffer)
  } else {
    bbox = turfBbox(feature)
  }
  const polygon = bboxPolygon(bbox)
  const tiles = tileCover.indexes(polygon.geometry, tileLimits)
  return tiles
}
