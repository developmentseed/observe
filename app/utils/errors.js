import serializeError from 'serialize-error'

export class ObserveError extends Error {
  /**
   * @param {String} code - default is 'ObserveError'
   * @param {String} message
   * @param {String / Error} extra - can either be a string or an Error instance
   */
  constructor (code = 'ObserveError', message, extra) {
    super(message)
    this.code = code
    this.retryable = false
    this.extra = extra
  }

  // Returns a JSON serializable representation of the error
  toJSON () {
    return {
      code: this.code,
      message: this.message,
      retryable: this.retryable,
      extra: this.extra instanceof Error ? serializeError(this.extra) : this.extra
    }
  }
}

export class UploadNetworkError extends ObserveError {
  constructor (message = 'Network error while uploading changes', extra) {
    super('UploadNetworkError', message, extra)
    this.retryable = true
  }
}

export class AuthError extends ObserveError {
  constructor (message = 'There was a problem authenticating the request', extra) {
    super('AuthError', message, extra)
    this.retryable = true
  }
}

export class ChangesetParseError extends ObserveError {
  constructor (message = 'Error parsing Changeset XML while uploading', extra) {
    super('ChangesetParseError', message, extra)
    this.retryable = false
  }
}

export class ChangesetNotFoundError extends ObserveError {
  constructor (message = 'Changeset not found', extra) {
    super('ChangesetNotFoundError', message, extra)
    this.retryable = false
  }
}

export class UnknownServerError extends ObserveError {
  constructor (message = 'An unknown server error occurred', extra) {
    super('UnknownServerError', message, extra)
    this.retryable = false
  }
}

export class VersionMismatchError extends ObserveError {
  constructor (message = 'This feature has been updated since you edited it', extra) {
    super('VersionMismatchError', message, extra)
    this.retryable = false
  }
}

export class OsmChangeError extends ObserveError {
  constructor (message = 'Error parsing OSM Change XML on the server', extra) {
    super('OsmChangeError', message, extra)
    this.retryable = false
  }
}

export class FeatureNotFoundError extends ObserveError {
  constructor (message = 'This feature is not found on OSM', extra) {
    super('FeatureNotFoundError', message, extra)
    this.retryable = false
  }
}

export class FeatureDeletedError extends ObserveError {
  constructor (message = 'This feature is deleted', extra) {
    super('FeatureDeletedError', message, extra)
    this.retryable = false
  }
}
