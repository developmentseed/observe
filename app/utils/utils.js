export function isNewId (id) {
  id = String(id)
  return id.startsWith('observe-')
}
