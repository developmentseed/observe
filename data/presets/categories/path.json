{
    "icon": "temaki-pedestrian",
    "name": "Paths",
    "members": [
        "highway/path",
        "highway/footway",
        "highway/footway/marked",
        "highway/footway/sidewalk",
        "highway/steps",
        "highway/cycleway",
        "highway/bridleway",
        "highway/pedestrian_line"
    ]
}
