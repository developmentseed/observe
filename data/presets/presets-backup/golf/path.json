{
    "icon": "maki-golf",
    "fields": [
        "name",
        "surface",
        "width",
        "structure",
        "incline"
    ],
    "geometry": [
        "line"
    ],
    "tags": {
        "golf": "path",
        "foot": "designated"
    },
    "reference": {
        "key": "golf",
        "value": "path"
    },
    "name": "Golf Walking Path"
}
