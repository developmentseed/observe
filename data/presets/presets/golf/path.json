{
    "icon": "maki-golf",
    "fields": [
        "name",
        "surface",
        "width",
        "structure",
        "incline"
    ],
    "geometry": [
        "line"
    ],
    "tags": {
        "golf": "path"
    },
    "addTags": {
        "golf": "path",
        "highway": "path",
        "foot": "designated"
    },
    "name": "Golf Walking Path"
}
