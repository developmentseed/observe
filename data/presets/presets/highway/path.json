{
    "icon": "iD-other-line",
    "fields": [
        "name",
        "surface",
        "width",
        "structure",
        "access",
        "incline"
    ],
    "moreFields": [
        "wheelchair",
        "lit",
        "smoothness",
        "trail_visibility",
        "sac_scale",
        "mtb/scale",
        "mtb/scale/uphill",
        "mtb/scale/imba",
        "horse_scale",
        "covered",
        "ref",
        "dog"
    ],
    "geometry": [
        "line"
    ],
    "terms": [
        "hike",
        "hiking",
        "trackway",
        "trail",
        "walk"
    ],
    "tags": {
        "highway": "path"
    },
    "name": "Path"
}
