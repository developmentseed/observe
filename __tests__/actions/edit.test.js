/* global fetch */
/* global describe */
/* global it */
/* global expect */
/* global console */
import { UploadNetworkError } from '../../app/utils/errors'
import { advanceTo } from 'jest-date-mock'
import configureStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import {
  addFeature,
  editFeature,
  deleteFeature,
  setAddPointGeometry,
  startEditUpload,
  editUploadFailed,
  editUploaded,
  uploadEdits
} from '../../app/actions/edit'
import { getFeature } from '../test-utils'
import fs from 'fs'
import path from 'path'
import state from '../fixtures/state.json'
import stateConflict from '../fixtures/state-conflict.json'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const mockFeature1 = getFeature('node/1')

// set mock date
advanceTo(1000)

describe('test sync edit actions', () => {
  it('should ADD_FEATURE', () => {
    const store = mockStore({})
    store.dispatch(addFeature(mockFeature1, 'test comment'))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'ADD_FEATURE',
      feature: mockFeature1,
      id: 'node/1',
      comment: 'test comment',
      timestamp: 1000
    })
  })

  it('should EDIT_FEATURE', () => {
    const store = mockStore({})
    const oldFeature = getFeature('node/1', { 'building': 'yes' })
    const newFeature = getFeature('node/1', { 'building': 'no' })
    store.dispatch(editFeature(oldFeature, newFeature, 'test comment'))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'EDIT_FEATURE',
      oldFeature,
      newFeature,
      id: 'node/1',
      comment: 'test comment',
      timestamp: 1000
    })
  })

  it('should DELETE_FEATURE', () => {
    const store = mockStore({})
    const feature = getFeature('node/1')
    store.dispatch(deleteFeature(feature, 'test comment'))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'DELETE_FEATURE',
      feature: feature,
      id: 'node/1',
      comment: 'test comment',
      timestamp: 1000
    })
  })

  it('should SET_ADD_POINT_GEOMETRY', () => {
    const store = mockStore({})
    store.dispatch(setAddPointGeometry([-1, 1]))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'SET_ADD_POINT_GEOMETRY',
      geometry: [-1, 1]
    })
  })

  it('should start edit upload', () => {
    const store = mockStore({})
    const edit = {
      oldFeature: null,
      newFeature: getFeature('node/1'),
      id: 'node/1'
    }
    store.dispatch(startEditUpload(edit))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'EDIT_UPLOAD_STARTED',
      edit: edit
    })
  })

  it('should handle upload failed', () => {
    const store = mockStore({})
    const edit = {
      oldFeature: null,
      newFeature: getFeature('node/1'),
      id: 'node/1'
    }
    const error = new UploadNetworkError()
    store.dispatch(editUploadFailed(edit, error))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'EDIT_UPLOAD_FAILED',
      edit,
      error: error.toJSON()
    })
  })

  it('should handle upload success', () => {
    const store = mockStore({})
    const edit = {
      oldFeature: null,
      newFeature: getFeature('node/1'),
      id: 'node/1'
    }
    const changesetId = '123'
    store.dispatch(editUploaded(edit, changesetId))
    const actions = store.getActions()
    expect(actions[0]).toEqual({
      type: 'EDIT_UPLOADED',
      edit,
      changesetId,
      timestamp: 1000
    })
  })
})

describe('test async edit actions', () => {
  it('should UPLOAD_EDITS', () => {
    const store = mockStore(state)
    const edit = {
      oldFeature: null,
      newFeature: { 'type': 'Feature', 'id': 'node/observe-vfxnxtmo20j', 'geometry': { 'type': 'Point', 'coordinates': [-77.02933996149397, 38.89503041477742] }, 'properties': { 'id': 'node/observe-vfxnxtmo20j', 'version': 1, 'name': 'Test', 'building': 'yes', 'icon': 'maki_marker' } },
      id: 'node/observe-vfxnxtmo20j'
    }
    const xmlDiffResponse = fs.readFileSync(path.join(__dirname, '../fixtures/osm-change-upload-response.xml'), 'utf-8')
    const osmResponseforTileUpdate = fs.readFileSync(path.join(__dirname, '../fixtures/osm-response-for-updated-tile.xml'), 'utf-8')

    fetch
      .once('1') // open changeset
      .once(xmlDiffResponse) // upload osm change
      .once({ status: 200 }) // close changeset
      .once(osmResponseforTileUpdate)

    return store.dispatch(uploadEdits([edit.id]))
      .then(() => {
        const actions = store.getActions()
        const types = []
        actions.forEach(action => {
          types.push(action.type)
        })
        expect(types).toStrictEqual([ 'EDIT_UPLOAD_STARTED',
          'EDIT_UPLOADED',
          'REQUESTED_TILE',
          'LOADING_DATA_FOR_TILE',
          'LOADED_DATA_FOR_TILE',
          'UPDATE_VISIBLE_BOUNDS',
          'REQUESTED_TILE',
          'REQUESTED_TILE' ])

        expect(actions).toMatchSnapshot()
      })
  })

  it('should raise VersionMismatchError', () => {
    const store = mockStore(stateConflict)
    const edit = {
      oldFeature: {
        'type': 'Feature',
        'id': 'node/1',
        'geometry': {
          'type': 'Point',
          'coordinates': [
            -77.02937206507221,
            38.89497324828185
          ]
        },
        'properties': {
          'id': 'node/1',
          'version': 1,
          'name': 'Test',
          'building': 'house',
          'icon': 'maki_marker'
        }
      },
      newFeature: {
        'type': 'Feature',
        'id': 'node/1',
        'geometry': {
          'type': 'Point',
          'coordinates': [
            -77.02937206507221,
            38.89497324828185
          ]
        },
        'properties': {
          'id': 'node/1',
          'version': 1,
          'name': 'Test',
          'building': 'yes',
          'icon': 'maki_marker'
        }
      },
      id: 'node/1'
    }

    fetch.resetMocks()
    fetch
      .once('1') // open changeset
      .once('Version mismatch: Provided 1, server had: 2 of Node 1', { status: 409 }) // upload osm change

    return store.dispatch(uploadEdits([edit.id]))
      .then(() => {
        const actions = store.getActions()
        expect(actions).toMatchSnapshot()
      })
  })
})
