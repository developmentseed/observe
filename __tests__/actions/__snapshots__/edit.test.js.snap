// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test async edit actions should UPLOAD_EDITS 1`] = `
Array [
  Object {
    "edit": Object {
      "comment": "Test",
      "errors": Array [],
      "id": "node/observe-vfxnxtmo20j",
      "newFeature": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.02937206507221,
            38.89497324828185,
          ],
          "type": "Point",
        },
        "id": "node/observe-vfxnxtmo20j",
        "properties": Object {
          "building": "yes",
          "icon": "maki_marker",
          "id": "node/observe-vfxnxtmo20j",
          "name": "Test",
          "version": 1,
        },
        "type": "Feature",
      },
      "oldFeature": null,
      "status": "EDIT_PENDING_STATUS",
      "timestamp": 1562325975841,
      "type": "create",
    },
    "type": "EDIT_UPLOAD_STARTED",
  },
  Object {
    "changesetId": "1",
    "edit": Object {
      "comment": "Test",
      "errors": Array [],
      "id": "node/observe-vfxnxtmo20j",
      "newFeature": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.02937206507221,
            38.89497324828185,
          ],
          "type": "Point",
        },
        "id": "node/observe-vfxnxtmo20j",
        "properties": Object {
          "building": "yes",
          "icon": "maki_marker",
          "id": "node/observe-vfxnxtmo20j",
          "name": "Test",
          "version": 1,
        },
        "type": "Feature",
      },
      "oldFeature": null,
      "status": "EDIT_PENDING_STATUS",
      "timestamp": 1562325975841,
      "type": "create",
    },
    "timestamp": 1000,
    "type": "EDIT_UPLOADED",
  },
  Object {
    "offline": false,
    "tile": "0320100322313223",
    "type": "REQUESTED_TILE",
  },
  Object {
    "tile": "0320100322313223",
    "type": "LOADING_DATA_FOR_TILE",
  },
  Object {
    "offline": false,
    "tile": "0320100322313223",
    "type": "LOADED_DATA_FOR_TILE",
  },
  Object {
    "type": "UPDATE_VISIBLE_BOUNDS",
    "visibleBounds": Array [
      Array [
        -77.02886915086447,
        38.8957164152211,
      ],
      Array [
        -77.02987497924005,
        38.8942300735456,
      ],
    ],
    "zoom": 18,
  },
  Object {
    "offline": false,
    "tile": "0320100322313221",
    "type": "REQUESTED_TILE",
  },
  Object {
    "offline": false,
    "tile": "0320100322313223",
    "type": "REQUESTED_TILE",
  },
]
`;

exports[`test async edit actions should raise VersionMismatchError 1`] = `
Array [
  Object {
    "edit": Object {
      "comment": "Test",
      "errors": Array [],
      "id": "node/1",
      "newFeature": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.02937206507221,
            38.89497324828185,
          ],
          "type": "Point",
        },
        "id": "node/1",
        "properties": Object {
          "building": "yes",
          "icon": "maki_marker",
          "id": "node/1",
          "name": "Test",
          "version": 1,
        },
        "type": "Feature",
      },
      "oldFeature": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.02937206507221,
            38.89497324828185,
          ],
          "type": "Point",
        },
        "id": "node/1",
        "properties": Object {
          "building": "house",
          "icon": "maki_marker",
          "id": "node/1",
          "name": "Test",
          "version": 1,
        },
        "type": "Feature",
      },
      "status": "EDIT_PENDING_STATUS",
      "timestamp": 1562325975841,
      "type": "modify",
    },
    "type": "EDIT_UPLOAD_STARTED",
  },
  Object {
    "level": "error",
    "message": "Upload failed",
    "type": "SET_NOTIFICATION",
  },
  Object {
    "edit": Object {
      "comment": "Test",
      "errors": Array [],
      "id": "node/1",
      "newFeature": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.02937206507221,
            38.89497324828185,
          ],
          "type": "Point",
        },
        "id": "node/1",
        "properties": Object {
          "building": "yes",
          "icon": "maki_marker",
          "id": "node/1",
          "name": "Test",
          "version": 1,
        },
        "type": "Feature",
      },
      "oldFeature": Object {
        "geometry": Object {
          "coordinates": Array [
            -77.02937206507221,
            38.89497324828185,
          ],
          "type": "Point",
        },
        "id": "node/1",
        "properties": Object {
          "building": "house",
          "icon": "maki_marker",
          "id": "node/1",
          "name": "Test",
          "version": 1,
        },
        "type": "Feature",
      },
      "status": "EDIT_PENDING_STATUS",
      "timestamp": 1562325975841,
      "type": "modify",
    },
    "error": Object {
      "code": "VersionMismatchError",
      "extra": "Version mismatch: Provided 1, server had: 2 of Node 1",
      "message": "This feature has been updated since you edited it",
      "retryable": false,
    },
    "type": "EDIT_UPLOAD_FAILED",
  },
]
`;
